tests/checkup_test.go

```diff
@@ -0,0 +1,254 @@
+ * limitations under the License.
+ *
+ * Copyright 2023 Red Hat, Inc.
+ *
+ */
+
+package tests
+
+import (
+	"context"
+	"time"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	. "github.com/onsi/gomega/gstruct"
+
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	testServiceAccountName                     = "rt-checkup-sa"
+	testKiagnoseConfigMapAccessRoleName        = "kiagnose-configmap-access"
+	testKiagnoseConfigMapAccessRoleBindingName = testKiagnoseConfigMapAccessRoleName
+	testConfigMapName                          = "rt-checkup-config"
+	testCheckupJobName                         = "rt-checkup"
+)
+
+var _ = Describe("Checkup execution", func() {
+	var (
+		configMap  *corev1.ConfigMap
+		checkupJob *batchv1.Job
+	)
+
+	BeforeEach(func() {
+		setupCheckupPermissions()
+
+		var err error
+		configMap = newConfigMap()
+		_, err = client.CoreV1().ConfigMaps(testNamespace).Create(context.Background(), configMap, metav1.CreateOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		DeferCleanup(func() {
+			err = client.CoreV1().ConfigMaps(testNamespace).Delete(context.Background(), testConfigMapName, metav1.DeleteOptions{})
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		checkupJob = newCheckupJob()
+		_, err = client.BatchV1().Jobs(testNamespace).Create(context.Background(), checkupJob, metav1.CreateOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		DeferCleanup(func() {
+			backgroundPropagationPolicy := metav1.DeletePropagationBackground
+			err = client.BatchV1().Jobs(testNamespace).Delete(
+				context.Background(),
+				testCheckupJobName,
+				metav1.DeleteOptions{PropagationPolicy: &backgroundPropagationPolicy},
+			)
+			Expect(err).NotTo(HaveOccurred())
+		})
+	})
+
+	It("should complete successfully", func() {
+		Eventually(getJobConditions, 5*time.Minute, 5*time.Second).Should(
+			ContainElement(MatchFields(IgnoreExtras, Fields{
+				"Type":   Equal(batchv1.JobComplete),
+				"Status": Equal(corev1.ConditionTrue),
+			})))
+
+		configMap, err := client.CoreV1().ConfigMaps(testNamespace).Get(context.Background(), testConfigMapName, metav1.GetOptions{})
+		Expect(err).NotTo(HaveOccurred())
+
+		Expect(configMap.Data).NotTo(BeNil())
+		Expect(configMap.Data["status.succeeded"]).To(Equal("true"))
+		Expect(configMap.Data["status.failureReason"]).To(BeEmpty())
+	})
+})
+
+func setupCheckupPermissions() {
+	var (
+		err                                error
+		checkupServiceAccount              *corev1.ServiceAccount
+		kiagnoseConfigMapAccessRole        *rbacv1.Role
+		kiagnoseConfigMapAccessRoleBinding *rbacv1.RoleBinding
+	)
+
+	checkupServiceAccount = newServiceAccount()
+	_, err = client.CoreV1().ServiceAccounts(testNamespace).Create(context.Background(), checkupServiceAccount, metav1.CreateOptions{})
+	Expect(err).NotTo(HaveOccurred())
+
+	DeferCleanup(func() {
```

Comment:
Interesting.. so this defer can work even if it is in a helper function?

Reply:
It cleans up the RoleBinding, Role and ServiceAccount.